### PR TITLE
Implementation of Coconsensus::finaliseCheckpoint

### LINKED ISSUE #199: Implement Coconsensus::finaliseCheckpoint()
1. Implement `Coconsensus::finaliseCheckpoint()`
    - Can be called by OriginProtocore or SelfProtocore
    - Updates the `blockchains` for the metachainId and number with the dynasty when this is called
    - Checks that the number is higher than the last finalised one
    - Proposal
    ```solidity
    function finalizeCheckpoint(
        bytes32 metachainId, 
        uint256 number, 
        bytes32 blockhash
    )
        onlyRunningProtocore(metachainId)
    ```

### PR DESCRIPTION
**Pull request by deepesh-kn**
This pull request contains the following:
1. Implementation of `Coconsensus::finaliseCheckpoint`.
2. Unit test for `Coconsensus::finaliseCheckpoint`.

fixes: #199 
blocked by: #244

### CODE REVIEW

File: contracts/coconsensus/Coconsensus.sol [hunk 1]
```diff
@@ -46,23 +47,310 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
...
+
+    /**
+     * @notice finaliseCheckpoint() function finalises a checkpoint at
+     *         a metachain.
+     *
+     * @param _metachainId A metachain id to finalise a checkpoint.
+     * @param _blockNumber A block number of a checkpoint.
+     * @param _blockHash A block hash of a checkpoint.
+     *
+     * \pre `_metachainId` is not 0.
```
Comment (pgev) [inline contracts/coconsensus/Coconsensus.sol]: Here we are checking that input parameters given by a protocore we trust are correct.
I am in favor of asserting more than less, however, not sure what are our policies for those cases hence raising this discussion. /cc @benjaminbollen

File: contracts/coconsensus/Coconsensus.sol [hunk 2]
```diff
@@ -46,23 +47,310 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
...
+     * \pre `_blockHash` is not 0.
+     * \pre `msg.sender` should be the protocore contract.
+     * \pre `_blockNumber` must be multiple of epoch length.
+     * \pre `_blockNumber` must be greater than the last finalized block number.
+     *
+     * \post Sets the `blockchains` mapping.
```
Comment (pgev) [inline contracts/coconsensus/Coconsensus.sol]: I want us to have `\post` and `\pre` conditions a little bit more verbose: in a way that a reader does not need to need a code and can understand them in the context of public/external interfaces of function/contract.

File: contracts/coconsensus/Coconsensus.sol [hunk 3]
```diff
@@ -46,23 +47,310 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
...
+            (_blockNumber % epochLength) == 0,
+            "Block number must be a checkpoint."
+        );
+
+        // Store the finalised block in the mapping.
+        Block storage finalisedBlock = blockchains[_metachainId][_blockNumber];
+        finalisedBlock.blockHash = _blockHash;
+        finalisedBlock.commitStatus = CheckpointCommitStatus.Finalized;
+
+        ProtocoreI selfProtocore = protocores[auxiliaryMetachainId];
```
Comment (pgev) [inline contracts/coconsensus/Coconsensus.sol]: This part should be updated as it has been agreed to keep `relative dynasty number` in coconsensus.

**Review by pgev — CHANGES_REQUESTED** [review]
@deepesh-kn thx for changes. Please, see my comments.

File: contracts/coconsensus/Coconsensus.sol (hunk 1 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol]: Sure, not making any changes for now.
@jayeshbairagi, please note this point for the discussion.

File: contracts/coconsensus/Coconsensus.sol (hunk 2 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol]: Updated.

File: contracts/coconsensus/Coconsensus.sol (hunk 3 above)
Comment (deepesh-kn) [reply contracts/coconsensus/Coconsensus.sol]: Now using `relativeSelfDynasty`

File: contracts/consensus/Coconsensus.sol [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+        uint256 lastFinalisedBlockNumber = blockTips[_metachainId];
+        require(
+            _blockNumber > lastFinalisedBlockNumber,
+            "The block number of the checkpoint must be greater than the block number of last finalised checkpoint."
+        );
+
+        // Check if the block number is multiple of epoch length.
+        ProtocoreI protocore = protocores[_metachainId];
+        uint256 epochLength = protocore.epochLength();
+        require(
```
Comment (pgev) [inline contracts/consensus/Coconsensus.sol]: I believe, this should be either removed (we trust implementation in Protocore) or changed to an assert.

File: contracts/consensus/Coconsensus.sol [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+            _blockHash != bytes32(0),
+            "Blockhash must not be null."
+        );
+
+        /*
+         * Check if the new block number is greater than the last
+         * finalised block number.
+         */
+        uint256 lastFinalisedBlockNumber = blockTips[_metachainId];
+        require(
```
Comment (pgev) [inline contracts/consensus/Coconsensus.sol]: Should this one be an assert?

File: contracts/consensus/Coconsensus.sol (hunk 4 above)
Comment (deepesh-kn) [reply contracts/consensus/Coconsensus.sol]: Removed.

File: contracts/consensus/Coconsensus.sol (hunk 5 above)
Comment (deepesh-kn) [reply contracts/consensus/Coconsensus.sol]: Changed to assert.

**Review by pgev — APPROVED** [review]

**Pull request merged**